graphql/util.go

```diff
@@ -1290,3 +1290,71 @@ func getAPISubscriptionsForProject(ctx context.Context, projectId string) ([]*re
 	}
 	return res, nil
 }
+
+// GroupProjects takes a list of projects and groups them by their repo. If onlyDefaultedToRepo is true,
+// it only groups projects that are defaulted to the repo
+func GroupProjects(projects []model.ProjectRef, onlyDefaultedToRepo bool) ([]*GroupedProjects, error) {
+	groupsMap := make(map[string][]*restModel.APIProjectRef)
+
+	for _, p := range projects {
+		groupName := fmt.Sprintf("%s/%s", p.Owner, p.Repo)
+
+		if onlyDefaultedToRepo && !p.UseRepoSettings {
+			groupName = ""
+		}
+
+		apiProjectRef := restModel.APIProjectRef{}
+		if err := apiProjectRef.BuildFromService(p); err != nil {
+			return nil, errors.Wrap(err, "error building APIProjectRef from service")
+		}
+
+		if projs, ok := groupsMap[groupName]; ok {
+			groupsMap[groupName] = append(projs, &apiProjectRef)
+		} else {
+			groupsMap[groupName] = []*restModel.APIProjectRef{&apiProjectRef}
+		}
+	}
+
+	groupsArr := []*GroupedProjects{}
+
+	for groupName, groupedProjects := range groupsMap {
+		gp := GroupedProjects{
+			Name:             &groupName, //deprecated
+			GroupDisplayName: groupName,
+			Projects:         groupedProjects,
+		}
+		project := groupedProjects[0]
+		if project.UseRepoSettings {
+			repoRefId := utility.FromStringPtr(project.RepoRefId)
+			repoRef, err := model.FindOneRepoRef(repoRefId)
+			if err != nil {
+				return nil, err
+			}
+
+			if repoRef == nil {
+				grip.Error(message.WrapError(err, message.Fields{
```

Comment:
I'm not sure on go conventions if err is nil here would that throw a nil pointer exception?

Reply:
we know that err isn't nil because of the check above. But either way, I don't think it will since we are not accessing err.Error.

Reply:
err is nil since the check above would have returned an error if error wasn't nil.

Reply:
ah yes, I read that backwards. I think it'll just log err as nil and log the rest of the fields. I'll double check before merging though.